src/state/cached_state.rs

```diff
@@ -383,16 +394,35 @@ impl<T: StateReader> State for CachedState<T> {
 
         // I: FETCHING FROM CACHE
         // deprecated contract classes dont have compiled class hashes, so we only have one case
-        if let Some(compiled_class) = self.contract_classes.get(class_hash) {
+        if let Some(compiled_class) = self.contract_class_cache_private.get_mut().get(class_hash) {
             return Ok(compiled_class.clone());
+        } else if let Some(compiled_class) =
+            self.contract_class_cache().get_contract_class(*class_hash)
+        {
+            self.contract_class_cache_private
+                .get_mut()
+                .insert(*class_hash, compiled_class.clone());
+            return Ok(compiled_class);
         }
 
         // I: CASM CONTRACT CLASS : CLASS_HASH
         if let Some(compiled_class_hash) =
             self.cache.class_hash_to_compiled_class_hash.get(class_hash)
         {
-            if let Some(casm_class) = self.contract_classes.get(compiled_class_hash) {
+            if let Some(casm_class) = self
+                .contract_class_cache_private
+                .get_mut()
+                .get(compiled_class_hash)
```

Comment:
Why do we need to `get_mut` here?

Reply:
We don't, but the alternative is borrowing it. Using [this method](https://doc.rust-lang.org/std/cell/struct.RefCell.html#method.get_mut) avoid the overhead that comes with borrow since by proving we have a mutable reference the compiler can be 100% sure that the refcell is not borrowed.

Reply:
Makes sense. Could you add that as a comment?

Reply:
Done. I've added it just to where your comment points, but there are other places where I'm using it too. Should I add it there too?

Reply:
Yes, that would be correct, but only when used in places where it's not used mutably (which is what makes it counter intuitive).